content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,357 @@
+the users and proved to be a success inside Google.
+
+The paper contains a number of general lessons like *Developer happiness is key* and *Crowdsource analysis development*
+that are nice, but I would rather highlight a few key takeaways instead, that seem to drive the rest of
+the technical decisions, responsible for the success of a new analysis platform.
+
+There are two main takeaways that drove the overall tooling design:
+
+### 1. Best way to **measure a success of analysis**
+> by number of bugs fixed (or prevented), not the number of issues identified
+
+This way of measuring a success has several notable implications:
+
+* If the tool that finds a bug also suggests a fix - it will be much more successful by this metric.
+  This by necessity constraints the scope of a possible analysis and a tooling required
+
+* It also means that repairing programs is important. For the discussion on tooling available for code
+  transformations see **Technical Details** section below.
+
+Learning such modifications from examples, instead of manual coding by engineers is also a bleeding edge
+research topic [https://github.com/KTH/learning4repair](https://github.com/KTH/learning4repair).
+
+### 2. Best way to **present the results of analysis**
+> Developers were more likely to fix bugs if presented with analysis results early and as part of their normal workflow
+
+This immediately implies that **reporting issues sooner is better.**
+
+It also leads to the conclusion that the best bet is to integrate checks either ***directly into compilers*** -
+familiar tools on whose feedback as errors and warnings developers are already relaying day to day. Or, if
+that is not possible, _**code review** is a good time_ for new changes — before they are committed to the
+version control system.
+
+Criteria that must hold for ***compile time*** checks:
+
+* 0 false positives
+
+* easy to understand, actionable and easy to fix
+
+* report issues affecting only correctness, rather than style or best practices
+
+Criteria that must hold for ***code review time*** checks:
+
+* <10% false positives
+
+* be understandable, actionable and easy to fix
+
+* have the potential for significant impact on code quality
+
+Last but not least, enabling any new *compile time* check can suddenly brake a developer working on existing
+code, which is clearly unacceptable. In order to avoid that, first **a tooling for large-scale code modifications**
+must be run over the whole codebase, before moving those checks to compile time for the rest of the company.
+
+And those are ClangMR and JavacFlume — projects that are only briefly mentioned in this insightful paper.
+
+*That is it, thank you for reading. We will post more on papers in this field soon.*
+
+{{% center %}} … {{% /center %}}
+
+Now I will take a liberty and cover a few technical details that were not in the scope of the original paper
```

Comment:
Would you be against splitting this into a second blog post?
I think the topic is by itself interesting.

Reply:
Moved to https://github.com/src-d/blog/issues/241